File: backend/src/main/java/com/smartcampost/backend/service/ParcelService.java
package com.smartcampost.backend.service;

import com.smartcampost.backend.dto.parcel.*;
import org.springframework.data.domain.Page;

import java.util.UUID;

public interface ParcelService {

    // US20: client creates a parcel (generates partial QR)
    ParcelResponse createParcel(CreateParcelRequest request);

    // Detail by ID (admin/staff or owner client)
    ParcelDetailResponse getParcelById(UUID parcelId);

    // Detail by tracking (client or admin)
    ParcelDetailResponse getParcelByTracking(String trackingRef);

    // List parcels for connected client
    Page<ParcelResponse> listMyParcels(int page, int size);

    // Global list (admin/staff)
    Page<ParcelResponse> listParcels(int page, int size);

    // US21: update status (requires ScanEvent)
    ParcelResponse updateParcelStatus(UUID parcelId, UpdateParcelStatusRequest request);

    // Accept parcel (CREATED -> ACCEPTED) - simple version
    ParcelResponse acceptParcel(UUID parcelId);

    // Accept parcel with full validation
    ParcelResponse acceptParcelWithValidation(UUID parcelId, AcceptParcelRequest request);

    // Change delivery option (AGENCY <-> HOME)
    ParcelResponse changeDeliveryOption(UUID parcelId, ChangeDeliveryOptionRequest request);

    // Update metadata (photo + comment)
    ParcelResponse updateParcelMetadata(UUID parcelId, UpdateParcelMetadataRequest request);

    // Pre-validation corrections (only when not locked)
    ParcelResponse correctParcelBeforeValidation(UUID parcelId, ParcelCorrectionRequest request);

    // Validate parcel and generate final QR code (GPS required)
    ParcelResponse validateAndLockParcel(UUID parcelId, Double latitude, Double longitude);

    // Admin-only exceptional override after lock (audited)
    ParcelResponse adminOverrideLockedParcel(UUID parcelId, AdminParcelOverrideRequest request);

    // Check if parcel can be corrected
    boolean canCorrectParcel(UUID parcelId);
}

File: smartcampost-frontend/src/services/parcels/parcelService.ts
import { api } from "../api/api";
import type { PageResponse } from "../../types/Common";
import type {
  ParcelResponse,
  ParcelDetailResponse,
  ParcelStatus,
  DeliveryOption,
  CreateParcelRequest,
} from "../../types/Parcel";

export interface UpdateParcelStatusRequest {
  status: ParcelStatus;
}

export interface ChangeDeliveryOptionRequest {
  deliveryOption: DeliveryOption;
}

export interface UpdateParcelMetadataRequest {
  comment?: string;
  photoUrl?: string;
}

export async function createParcel(payload: CreateParcelRequest) {
  const { data } = await api.post<ParcelResponse>("/api/parcels", payload);
  return data;
}

export async function listMyParcels(page = 0, size = 20) {
  const { data } = await api.get<PageResponse<ParcelResponse>>("/api/parcels/me", {
    params: { page, size },
  });
  return data;
}

export async function listParcels(page = 0, size = 20) {
  const { data } = await api.get<PageResponse<ParcelResponse>>("/api/parcels", {
    params: { page, size },
  });
  return data;
}

export async function getParcelById(parcelId: string) {
  const { data } = await api.get<ParcelDetailResponse>(`/api/parcels/${parcelId}`);
  return data;
}

export async function getParcelByTracking(trackingRef: string) {
  const { data } = await api.get<ParcelDetailResponse>(`/api/parcels/tracking/${trackingRef}`);
  return data;
}

export async function updateParcelStatus(parcelId: string, payload: UpdateParcelStatusRequest) {
  const { data } = await api.patch<ParcelResponse>(`/api/parcels/${parcelId}/status`, payload);
  return data;
}

export async function acceptParcel(parcelId: string) {
  const { data } = await api.patch<ParcelResponse>(`/api/parcels/${parcelId}/accept`);
  return data;
}

export async function changeDeliveryOption(parcelId: string, payload: ChangeDeliveryOptionRequest) {
  const { data } = await api.patch<ParcelResponse>(`/api/parcels/${parcelId}/delivery-option`, payload);
  return data;
}

export async function updateParcelMetadata(parcelId: string, payload: UpdateParcelMetadataRequest) {
  const { data } = await api.patch<ParcelResponse>(`/api/parcels/${parcelId}/metadata`, payload);
  return data;
}
